llvm/lib/CodeGen/FPBuiltinFnSelection.cpp

```diff
@@ -0,0 +1,178 @@
+//===- FPBuiltinFnSelection.cpp - Pre-ISel intrinsic lowering pass --------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This pass implements alternate math library implementation selection for
+// llvm.fpbuiltin.* intrinsics.
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/CodeGen/FPBuiltinFnSelection.h"
+#include "llvm/Analysis/TargetLibraryInfo.h"
+#include "llvm/CodeGen/Passes.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/IntrinsicInst.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Support/FormatVariadic.h"
+
+using namespace llvm;
+
+#define DEBUG_TYPE "fpbuiltin-fn-selection"
+
+static bool replaceWithAltMathFunction(FPBuiltinIntrinsic &BuiltinCall,
```

Comment:
Just to understand, are we going to lose all fpbuiltin- meta data in this transformation? Thanks

Reply:
Yes, that is my intention. After an implementation has been selected, the function being called should meet the specified requirements and should not be replaced. If we kept the attributes, it would mean that later passes (specifically instruction selection) would need to know what to do with them.